incubator/hnc/pkg/forest/forest.go

```diff
@@ -205,3 +220,83 @@ func (ns *Namespace) IsAncestor(other *Namespace) bool {
 	}
 	return ns.parent.IsAncestor(other)
 }
+
+// ClearConditions clears local conditions in the namespace.
+func (ns *Namespace) ClearConditions(key string) {
+	delete(ns.conditions, key)
+}
+
+// SetCondition adds a condition into the list of conditions for key string, returning
+// true if it does not exist previously.
+func (ns *Namespace) SetCondition(key string, code tenancy.Code, msg string) bool {
+	oldConditions := ns.conditions[key]
+	for _, condition := range oldConditions {
+		if condition.code == code && condition.msg == msg {
+			return false
+		}
+	}
+
+	ns.conditions[key] = append(oldConditions, condition{code: code, msg: msg})
+	return true
+}
+
+// Conditions returns a list of conditions in the namespace.
+// It converts map[string][]condition into []Condition.
+func (ns *Namespace) Conditions() []tenancy.Condition {
+	return flatten(ns.convertConditions())
+}
+
+// convertConditions converts string -> condition{code, msg} map into condition{code, msg} -> affected map.
+func (ns *Namespace) convertConditions() map[tenancy.Code]map[string][]tenancy.AffectedObject {
+	converted := map[tenancy.Code]map[string][]tenancy.AffectedObject{}
+	for key, conditions := range ns.conditions {
+		for _, condition := range conditions {
+			if affected, ok := converted[condition.code][condition.msg]; !ok {
+				converted[condition.code] = map[string][]tenancy.AffectedObject{}
+				converted[condition.code][condition.msg] = []tenancy.AffectedObject{getAffectedObject(key)}
+			} else {
+				converted[condition.code][condition.msg] = append(affected, getAffectedObject(key))
+			}
+		}
+	}
+	return converted
+}
+
+// flatten flattens condition{code, msg} -> affected map into a list of condition{code, msg, []affected}.
+func flatten(m map[tenancy.Code]map[string][]tenancy.AffectedObject) []tenancy.Condition {
+	flattened := []tenancy.Condition{}
+	for code, msgAffected := range m {
+		for msg, affected := range msgAffected {
+			flattened = append(flattened, tenancy.Condition{Code: code, Msg: msg, Affects: affected})
+		}
+	}
+	if len(flattened) > 0 {
+		return flattened
+	}
+	return nil
+}
+
+// getAffectedObject gets AffectedObject from a namespace or a string of form group/version/kind/namespace/name
+func getAffectedObject(gvknn string) tenancy.AffectedObject {
+	if gvknn == Local {
+		return tenancy.AffectedObject{}
```

Comment:
I think `return nil` would be better here?

Reply:
Currently this func returns a copy of the AffectedObject struct. I changed it into returning a pointer and an error (*tenancy.AffectedObject, error). It returns error if the key is not valid to get an AffectedObject.

Reply:
As discussed offline, don't return an error here since we don't have any way to handle it. Instead, maybe return an `AffectedObject` with the namespace set to "INVALID OBJECT" (which is not a valid namespace name) and the name set to the key. Also pass in the `log` so you can log an error.

Also, don't return a pointer like I suggested - I misread this and thought it was a slice, but it's a single object.